### ISSUE TITLE
Twig Templates

### ISSUE THREAD

**Issue by Fruchuxs** [contao/core-bundle#1215]
In alten Feature-Requests und Co habe ich des öfteren gelesen, dass es wohl geplant ist, die Templates mit Twig zu realisieren. Zurück in der Zukunft bei Contao 4, sehe ich immer noch diese "ätzenden" PHP-Templates.
Ist das noch geplant, oder eher nicht?

**Comment by aschempp** [comment]
Grundsätzlich kann Twig problemlos von Erweiterungen verwendet werden. Das primäre Problem mit Twig ist, dass Contao Input-Encoding betreibt, und Twig automatisch Output-Encoding macht. Das zweite Problem ist selbstverständlich dass wir nicht alle Templates einfach von heute auf morgen ersetzen können, denn dann wären ja alle angepassten Templates hinfällig. Aber wir hätten schon gerne Twig…

**Comment by m-vo** [comment]
>  Das zweite Problem ist selbstverständlich dass wir nicht alle Templates einfach von heute auf morgen ersetzen können, denn dann wären ja alle angepassten Templates hinfällig. 

Wenn die Templates funktional identisch wären, und der PHP-Fallback immer noch funktionieren würde könnte das doch mit einem Minor-Release machbar sein? (vielleicht ein bisschen unintuitiv für den Benutzer, der sich zuvor die Sachen im PHP-Template abgeschaut hat, aber ggf. vertretbar) Oder übersehe ich dabei noch andere Probleme?

**Comment by backbone87** [comment]
Löst nicht das Problem von Input encoding vs Output encoding

**Comment by m-vo** [comment]
Dass Output Encoding auf lange Sicht die bessere Wahl ist, ist vermutlich Konsens. :-) Allerdings sind das meiner Meinung nach zwei unabhängige Probleme. 

Wir könnten z.B. in den Templates ``{{ something | raw }}`` benutzen und das ``| raw`` irgendwann droppen, wenn wir von Input auf Output Encoding umsteigen. Dann wären wir trotzdem schon ein gutes Stück weiter.

**Comment by Fruchuxs** [comment]
Würde man Daten unkodiert in der DB abspeichern, kann man auch vor dem Setzen der Daten in den *.html5-Templates die Kodierung dort wieder vorschalten und das Template-System zwei-gleisig fahren. Die Abwärtskompatibilität wäre wieder gegeben. Wobei ich zugeben muss, dass ich nicht genau weiß, wie contao Daten intern handhabt und wie aufwendig das wäre. 

Nur ist Twig als Template-System in contao sicher eine Bereicherung. Dieses HTML/PHP-Mischmasch ist ziemlich schwer zu lesen. Als Minor-Release würde es ja schon reichen, wenn contao *.twig-Templates durch den Twig-Parser jagt.

**Comment by backbone87** [comment]
ja man kann das output encoding in twig ausschalten, aber dann bist du im prinzip da wo man jetzt ist: du kannst twig templates schreiben und sie mit dem twig service rendern. gut das ist noch nicht in CEs & Co integriert, aber das ist ja mehr ne fleißaufgabe.

aus meiner sicht gibt es kein pfad für eine softmigration von input zu output encoding. einfach in der DB rumschreiben funktioniert vll in kleineren systemen, wo man von hand ran gehen kann um ein paar dutzend sachen per hand zu korrigieren. aber in größeren systemen wird das ein horror.

**Comment by Fruchuxs** [comment]
Wer spricht von "in der DB rumschreiben"? Wie Daten gespeichert werden, ist einer Datenbank per se egal. Die Frage ist nur, wie die Systeme, die die Daten wieder auslesen, diese erwarten.

**Comment by aschempp** [comment]
Input Encoding lässt sich praktisch nicht rückgängig machen, daher wird es schwierig das mit Twig zu benutzen.

**Comment by Fruchuxs** [comment]
Ich habe tatsächlich nur Neuinstallationen im Auge gehabt. Da habt Ihr einfach den besseren Überblick.
Allerdings ist es auch egal mit dem Encoding, da man bei Twig auch global das autoescaping deaktivieren kann (was allerdings nur Sinn macht, wenn man für das Template-System von Contao selbst eine eigene Instanz verwendet). 
Dann noch ein Schalter "ALTES SYSTEM" "ja|nein". Dann liegt die Verantwortung bei Verwender. Oder übersehe ich etwas?

**Comment by Toflar** [comment]
Ja, viele Extension Entwickler verlassen sich darauf, dass die DB encodeten Input liefert und geben ihn wieder in der gleichen Form aus. Auch wenn du es im System umstellst, werden alle Extensions nicht automatisch output encoding betreiben und du hast einen Scherbenhaufen voller Sicherheitslücken.

**Comment by Fruchuxs** [comment]
Deswegen hab ich ja geschrieben, Encoding von Twig ausschalten und das Verhalten von Contao beibehalten .. ;) oder zumindest impliziert.
Und "schon" hat man Twig-Templates anstelle von PHP-Templates und es muss nichts weiter an den Gegebenheiten von Contao modifiziert werden.

Es geht mir primär darum ein bewährtes Template System nutzen zu können, was vor allem die Lesbarkeit erhöht. Twig konnte ich schon vielen Leute nahe bringen die nur HTML können, bei diesen PHP-Templates hört es aber auf.

**Comment by andyKempf** [comment]
Statt den bisherigen PHP-Templates dann Twig als Template-Engine einzusetzen wäre ein Traum!

**Comment by andyKempf** [comment]
Ist schon bekannt, ob in naher oder ferner Zukunft Frontend-Templates mit Twig ausgeliefert werden? Das wäre einfach großartig für eine professionelle Frontend-Entwicklung. Auch im Zusammenhang mit dem neuen Symfony 4.

**Comment by Toflar** [comment]
In ferner Zukunft, vielleicht. Wie hier erklärt, in Contao 4 wird es wohl nie der Fall sein.

**Comment by fritzmg** [comment]
@m-vo und ich haben damit am Contao Camp nochmal ein wenig experimentiert. Als PoC hatte ich die `fe_page` komplett über twig aufgebaut, was dann auch funktioniert hat. Ob es das Wert ist, alle notwendigen Änderungen in das aktuelle Template System von Contao einzubauen und ob wir das wirklich wollen ist natürlich eine andere Frage ;). Möglich wäre es.

**Comment by andyKempf** [comment]
@fritzmg @m-vo das hört sich großartig an! 👍🏼 Wie aufwändig waren die Änderungen? Könnte ich mehr darüber erfahren? 🙂

Ansonsten verstehe ich schon, dass eine Umstellung auf Twig ein gewagter Schritt sein könnte, da sich Contao ja auch bei „kleineren“ Webdesignern großer Beliebtheit erfreut, diese sich aber bestimmt in der Vergangenheit nie mit einer Template-Engine wie Twig beschäftigt haben. Andererseits hält Contao als Bundle immer mehr Einzug in das Symfony-Framework, in dem es „State of the Art“ ist, Twig als Template-Engine im Frontend einzusetzen. 

Ich denke in diesem Zusammenhang auch, dass sich die Zielgruppe von Contao leicht verändern wird. Schließlich erfreut sich Symfony bei professionellen Webentwicklern großer Beliebtheit und Contao erhält als Bundle nochmal eine deutlich breitere Öffentlichkeit. Ähnlich läuft es auch bei Systemen wie Drupal oder Shopware.

**Comment by bytehead** [comment]
@andyKempf das Problem liegt ja auch nicht bei den "kleineren" Webdesignern oder anderen Zielgruppen - es ist technischer Natur: https://github.com/contao/core-bundle/issues/1215#issuecomment-348753597

**Comment by fritzmg** [comment]
> @fritzmg @m-vo das hört sich großartig an! 👍🏼 Wie aufwändig waren die Änderungen? Könnte ich mehr darüber erfahren? 🙂

Es war nur ein quick & dirty PoC. Die nötigen Änderungen wären schon etwas umfangreicher, damit das mit dem automatischen Template Resolving von Contao funktionieren würde beispielsweise. Und dann bleibt immer noch das von @bytehead erwähnte Problem.

(Ich kann mich aber nicht mehr erinnern, wie wir das gehandhabt haben - leider habe ich den Code schon verworfen 😬 @m-vo kannst du dich noch erinnern? 😉)

**Comment by m-vo** [comment]
Wir hatten soweit ich mich erinnere alles im Template Renderer hardcoded (fe_page.twig statt original rendern, Twig als Engine). Und dann die existierenden Funktionen und Variablen gebunden (recht straight forward). Für eine ordentliche Implementierung bräuchte es neben den angesprochenen Problemen (s. `|raw` modifier) einiges an Hirnschmalz/Arbeit für Template Selektion / BC Layer / Vererbung + Mischeinsatz alte/neue Templates / ...
